File: nestjs-api/src/services/user.service.ts
import { Injectable, NotFoundException } from '@nestjs/common';
import { InjectRepository } from '@nestjs/typeorm';
import { Repository } from 'typeorm';
import { User } from '../models/user.entity';
import { CreateUserDto, UpdateUserDto } from '../models/dto/create-user.dto';

@Injectable()
export class UserService {
  constructor(
    @InjectRepository(User)
    private userRepository: Repository<User>,
  ) {}

  async findAll(limit: number = 10, offset: number = 0): Promise<User[]> {
    return this.userRepository.find({
      take: limit,
      skip: offset,
      order: { id: 'ASC' },
    });
  }

  async findOne(id: number): Promise<User> {
    const user = await this.userRepository.findOne({ where: { id } });
    if (!user) {
      throw new NotFoundException(`User with ID ${id} not found`);
    }
    return user;
  }

  async create(createUserDto: CreateUserDto): Promise<User> {
    const user = this.userRepository.create(createUserDto);
    return this.userRepository.save(user);
  }

  async update(id: number, updateUserDto: UpdateUserDto): Promise<User> {
    const user = await this.findOne(id);
    Object.assign(user, updateUserDto);
    return this.userRepository.save(user);
  }

  async remove(id: number): Promise<{ message: string }> {
    const user = await this.findOne(id);
    await this.userRepository.remove(user);
    return { message: 'User deleted successfully' };
  }

  // Complex queries
  async getUserOrderSummary(limit: number = 10, offset: number = 0) {
    return this.userRepository.createQueryBuilder('u')
      .leftJoin('u.orders', 'o')
      .select('u.id', 'user_id')
      .addSelect('u.name', 'user_name')
      .addSelect('u.email', 'user_email')
      .addSelect('COUNT(o.id)', 'total_orders')
      .addSelect('COALESCE(SUM(o.total_amount), 0)', 'total_amount')
      .addSelect('COALESCE(AVG(o.total_amount), 0)', 'average_order')
      .addSelect('COALESCE(MAX(o.order_date), \'1970-01-01\')', 'last_order')
      .groupBy('u.id')
      .addGroupBy('u.name')
      .addGroupBy('u.email')
      .orderBy('total_amount', 'DESC')
      .limit(limit)
      .offset(offset)
      .getRawMany();
  }
} 

File: python-api/services/user_service.py
from sqlalchemy.orm import Session
from sqlalchemy import text
from models.models import User
from models.schemas import UserCreate, UserUpdate
from typing import List, Optional

class UserService:
    def __init__(self, db: Session):
        self.db = db

    def get_users(self, limit: int = 10, offset: int = 0) -> List[User]:
        return self.db.query(User).offset(offset).limit(limit).all()

    def get_user(self, user_id: int) -> Optional[User]:
        return self.db.query(User).filter(User.id == user_id).first()

    def create_user(self, user_data: UserCreate) -> User:
        db_user = User(**user_data.dict())
        self.db.add(db_user)
        self.db.commit()
        self.db.refresh(db_user)
        return db_user

    def update_user(self, user_id: int, user_data: UserUpdate) -> Optional[User]:
        db_user = self.get_user(user_id)
        if not db_user:
            return None
        
        update_data = user_data.dict(exclude_unset=True)
        for field, value in update_data.items():
            setattr(db_user, field, value)
        
        self.db.commit()
        self.db.refresh(db_user)
        return db_user

    def delete_user(self, user_id: int) -> bool:
        db_user = self.get_user(user_id)
        if not db_user:
            return False
        
        self.db.delete(db_user)
        self.db.commit()
        return True

    def get_user_order_summary(self, limit: int = 10, offset: int = 0):
        query = text("""
            SELECT u.id as user_id, u.name as user_name, u.email as user_email, 
                   COUNT(o.id) as total_orders,
                   COALESCE(SUM(o.total_amount), 0) as total_amount,
                   COALESCE(AVG(o.total_amount), 0) as average_order,
                   COALESCE(MAX(o.order_date), '1970-01-01'::timestamp) as last_order
            FROM users u
            LEFT JOIN orders o ON u.id = o.user_id
            GROUP BY u.id, u.name, u.email
            ORDER BY total_amount DESC
            LIMIT :limit OFFSET :offset
        """)
        
        result = self.db.execute(query, {"limit": limit, "offset": offset})
        return result.fetchall() 